src/jit/lsra.cpp

```diff
@@ -6937,7 +6930,26 @@ LinearScan::resolveLocalRef(GenTreePtr treeNode, RefPosition * currentRefPositio
         if (treeNode != nullptr)
         {
             treeNode->gtFlags |= GTF_SPILLED;
-            if (spillAfter) treeNode->gtFlags |= GTF_SPILL;
+            if (spillAfter)
+            {
+                if (currentRefPosition->AllocateIfProfitable())
+                {
+                    // This is a use of lclVar that is flagged as reg-optional
+                    // by lower/codegen and marked for both reload and spillAfter.
+                    // In this case we can avoid unnecessary reload and spill
+                    // by setting reg on lclVar to REG_STK and reg on tree node
+                    // to REG_NA.  Codegen will generate the code by considering
+                    // it as a contained memory operand.
+                    //
+                    // Note that varDsc->lvRegNum is already to REG_STK above.
+                    interval->physReg = REG_NA;
+                    treeNode->gtRegNum = REG_NA;
+                }
+                else
+                {
+                    treeNode->gtFlags |= GTF_SPILL;
+                }
+            }
         }
```

Comment:
I'm not sure that it matters, but I think it would be cleaner to clear the `GTF_SPILLED` flag in the then clause, since it is not actually going to be reloaded. I **think** that's what currently happens when the `AllocateIfProfitable` ones are not allocated in the first place (i.e. they do not get the `GTF_SPILLED` flag set).